framework/src/variables/MooseVariableDataFV.C

```diff
@@ -802,13 +804,30 @@ MooseVariableDataFV<OutputType>::computeAD(const unsigned int num_dofs, const un
   if (_need_ad_u_dot)
     _ad_u_dot[0] = _ad_zero;
 
+  // Limit the number of virtual calls
+  bool currently_computing_jacobian = _subproblem.currentlyComputingJacobian();
+
   for (unsigned int i = 0; i < num_dofs; i++)
   {
     _ad_dof_values[i] = (*_sys.currentSolution())(_dof_indices[i]);
 
-    // NOTE!  You have to do this AFTER setting the value!
-    if (_var.kind() == Moose::VAR_NONLINEAR)
+    // NOTE!  You have to do this AFTER setting the value! Also note that we get inside this method
+    // even during residual evaluation now. We usually stop ourselves from doing needless AD
+    // operations by setting the `DualNumber::do_derivatives` member to `false` during residual
+    // evaluation, but below we are bypassing `DualNumber` and working directly with its derivative
+    // container. So here we explicitly query whether we are doing residual/Jacobian evaluation to
+    // stop ourselves from doing an unnecessary op, and an op that is actually dangerous for global
+    // indexing. This is dangerous because when evaluating the residual the above setting of the
+    // `DualNumber::value` *doesn't* clear the derivative container like it does during Jacobian
+    // evaluation. So you will gradually build a derivative container containing a variable's entire
+    // set of dofs if you do the op below, and you'll quickly exceed the size of the underlying
+    // derivative storage container
+    if (_var.kind() == Moose::VAR_NONLINEAR && currently_computing_jacobian)
+#ifdef MOOSE_GLOBAL_AD_INDEXING
+      Moose::derivInsert(_ad_dof_values[i].derivatives(), _dof_indices[i], 1.);
```

Comment:
Would it be better to include some sort of explicit reset for the derivative values (while retaining the computing jacobian check) - just to make things less dangerous?  After reading this long comment, I'm still not sure I understand how derivatives build up.  Wouldn't the insert operation just see the same global dof index entry and re-insert the 1's to overwrite the existing entries?  Is this problem just because the derivInsert function is "dumb" and doesn't try to match up the inserted indices with existing indices?

Reply:
There is only one data object per variable. This routine will get called for every element, so by the end of the computeResidual routine, we will have tried to insert every dof index for a given variable. So in the case of CONSTANT MONOMIALS, if I have more than 50 elements, then I'm screwed.

I could explicitly clear the derivative vector when doing residual evaluations when I do the `_ad_dof_values` assignment since that is what is actually done during Jacobian evaluation (behind the scenes in MetaPhysicL)

Reply:
If for some reason people are doing PJFNK with FV with the default configuration (non-sparse, local indexing) explicit zeroing of the derivative container will be expensive. Perhaps I will just explicitly zero when we are doing global indexing